pkg/virt-launcher/virtwrap/api/converter.go

```diff
@@ -1014,27 +1010,64 @@ func Convert_v1_VirtualMachine_To_api_Domain(vmi *v1.VirtualMachineInstance, dom
 	return nil
 }
 
-func calculateRequestedVCPUs(vmi *v1.VirtualMachineInstance) uint32 {
-	cores := uint32(0)
-	if vmi.Spec.Domain.CPU != nil && vmi.Spec.Domain.CPU.Cores != 0 {
+func getCPUTopology(vmi *v1.VirtualMachineInstance) *CPUTopology {
+	return &CPUTopology{
+		Sockets: calculateRequestedSockets(vmi),
+		Cores:   calculateRequestedCPUs(vmi),
+		Threads: calculateRequestedThreads(vmi),
+	}
+}
+
+func calculateRequestedSockets(vmi *v1.VirtualMachineInstance) uint32 {
+	sockets := uint32(1)
+	cpu := vmi.Spec.Domain.CPU
+	if cpu.Sockets != 0 {
+		return cpu.Sockets
+	}
+
+	if cpu.Cores == 0 && cpu.Threads == 0 {
+		//if cores, sockets, threads are not set, take value from domain resources request or limits and
+		//set value into sockets, which have best performance (https://bugzilla.redhat.com/show_bug.cgi?id=1653453)
+		resources := vmi.Spec.Domain.Resources
+		if cpuRequests, ok := resources.Requests[k8sv1.ResourceCPU]; ok {
+			sockets = uint32(cpuRequests.Value())
+		} else if cpuLimit, ok := resources.Limits[k8sv1.ResourceCPU]; ok {
+			sockets = uint32(cpuLimit.Value())
+		}
+	}
+	return sockets
+}
+
+func calculateRequestedThreads(vmi *v1.VirtualMachineInstance) uint32 {
+	threads := uint32(1)
+	if vmi.Spec.Domain.CPU.Threads != 0 {
+		return vmi.Spec.Domain.CPU.Threads
+	}
+	return threads
+}
+
+func calculateRequestedVCPUs(cpuTopology *CPUTopology) uint32 {
+	return cpuTopology.Cores * cpuTopology.Sockets * cpuTopology.Threads
+}
+
+func calculateRequestedCPUs(vmi *v1.VirtualMachineInstance) uint32 {
```

Comment:
Shouldn't this be deleted now?

Reply:
No, because you still need to get cores from the vmi.